### PR TITLE
new call interface

### PR DESCRIPTION
**Pull request by blampe**
@breerly @abhinav @junchaowu

### CODE REVIEW

**Comment by breerly** [comment]
hold on to your butts

File: CHANGES.rst [hunk 1]
```diff
@@ -2,10 +2,19 @@ Changelog
...
 
-0.14.1 (unreleased)
+0.15.0 (unreleased)
 -------------------
 
-- Nothing changed yet.
+- Introduced new top level ``tchannel.TChannel`` object, with new request methods
+  ``call``, ``raw``, ``json``, and ``thrift``. This will eventually replace the
+  akward ``request`` / ``send`` calling pattern.
+- Introduced ``tchannel.from_thrift_module`` function for creating a request builder
```
Comment (abhinav) [inline CHANGES.rst]: I thought we were still trying to decide a better name for this?

File: CHANGES.rst [hunk 2]
```diff
@@ -2,10 +2,19 @@ Changelog
...
+  to be used with the ``tchannel.TChannel.thrift`` function.
+- Introduced new simplified examples under the ``examples/simple`` directory, moved
+  the Guide's examples to ``examples/guide``, and deleted the remaining examples.
+- Added ThriftTest.thrift and generated Thrift code to ``tchannel.testing.data`` for
+  use with examples and playing around with TChannel.
```
Comment (abhinav) [inline CHANGES.rst]: Not sure what value this serves to users.

File: docs/api.rst [hunk 3]
```diff
@@ -5,7 +5,16 @@ API Documentation
...
+.. autoclass:: tchannel.TChannel
+    :members:
+
+.. autoclass:: tchannel.schemes.RawArgScheme
+    :members:
+
+.. autoclass:: tchannel.schemes.JsonArgScheme
+    :members:
+
+.. autoclass:: tchannel.schemes.ThriftArgScheme
```
Comment (abhinav) [inline docs/api.rst]: People don't use scheme objects directly. Why expose them in the reference?

File: tchannel/response.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+)
+
+from . import transport
+
+__all__ = ['Response']
+
+
+class Response(object):
+
+    # TODO add __slots__
```
Comment (abhinav) [inline tchannel/response.py]: Docs

File: tchannel/retry.py [hunk 5]
```diff
@@ -0,0 +1,12 @@
+from __future__ import (
+    absolute_import, division, print_function, unicode_literals
+)
+
+NEVER = 'n'
+CONNECTION_ERROR = 'c'
+TIMEOUT = 't'
+CONNECTION_ERROR_AND_TIMEOUT = 'ct'
+DEFAULT = CONNECTION_ERROR
+
```
Comment (abhinav) [inline tchannel/retry.py]: Docs on all.

Use the following format to get sphinx to pick it up.

    #: docstring
    CONSTANT = value

File: tchannel/schemes/json.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+            )
+
+        :param string service:
+            Name of the service to call.
+        :param string endpoint:
+            Endpoint to call on service.
+        :param string body:
+            A raw body to provide to the endpoint.
+        :param string headers:
+            A raw headers block to provide to the endpoint.
```
Comment (abhinav) [inline tchannel/schemes/json.py]: 'raw'?

File: tchannel/schemes/json.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+        :param int timeout:
+            How long to wait before raising a ``TimeoutError`` - this
+            defaults to ``tchannel.glossary.DEFAULT_TIMEOUT``.
```
Comment (abhinav) [inline tchannel/schemes/json.py]: no unit specified?

File: tchannel/schemes/json.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+        :param string retry_on:
+            What events to retry on - valid values can be found in
+            ``tchannel.retry``.
+        :param string retry_limit:
+            How many times to retry before
```
Comment (abhinav) [inline tchannel/schemes/json.py]: retry limit is probably an integer, not a string

File: tchannel/schemes/json.py [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+        :return Response:
+        """
+
+        # TODO should we not default these?
+        if headers is None:
+            headers = {}
+
+        # TODO dont default?
+        if body is None:
+            body = {}
```
Comment (abhinav) [inline tchannel/schemes/json.py]: Don't default. Null is a valid JSON value.

File: tchannel/schemes/thrift.py [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+        for exc_spec in result_spec[1:]:
+            exc = getattr(body, exc_spec[2])
+            if exc is not None:
+                raise exc
+
+        # success - non-void
+        if len(result_spec) >= 1 and result_spec[0] is not None:
+            response.body = body.success
+            raise gen.Return(response)
+
```
Comment (abhinav) [inline tchannel/schemes/thrift.py]: if `result_spec[0]` was not None and no value was returned, that's a user-error and an exception needs to be raised

File: tchannel/schemes/json.py (hunk 9 above)
Comment (breerly) [inline tchannel/schemes/json.py]: this failed before adding this - haven't had time to dig deep into it but this works for now

File: tchannel/schemes/json.py (hunk 8 above)
Comment (breerly) [inline tchannel/schemes/json.py]: dis is true

File: tchannel/schemes/json.py (hunk 7 above)
Comment (breerly) [inline tchannel/schemes/json.py]: hmm? I think what I want to do is expand the value at DEFAULT_TIMEOUT, I'm pretty sure sphinx will let us do that I just haven't before

File: tchannel/schemes/json.py (hunk 6 above)
Comment (breerly) [inline tchannel/schemes/json.py]: I more or less a "free-form headers block"... suggestions on wording are welcome

File: CHANGES.rst (hunk 1 above)
Comment (breerly) [inline CHANGES.rst]: We were

File: tchannel/schemes/thrift.py (hunk 10 above)
Comment (breerly) [inline tchannel/schemes/thrift.py]: Will make task now.

**Pull request merged**